src/code/padmgr.c

```diff
@@ -1,58 +1,797 @@
+ */
+void PadMgr_GetInputNoLock(Input* inputs, s32 gameRequest) {
+    s32 i;
+    Input* input = &sPadMgrInstance->inputs[0];
+    Input* inputOut = &inputs[0];
+    s32 buttonDiff;
+
+    for (i = 0; i < MAXCONTROLLERS; i++) {
+        if (gameRequest) {
+            // Copy inputs as-is, press and rel are calculated prior in `PadMgr_UpdateInputs`
+            *inputOut = *input;
+            // Zero parts of the press and rel inputs in the polled inputs so they are not read more than once
+            input->press.button = 0;
+            input->rel.button = 0;
+        } else {
+            // Take as the previous inputs the inputs that are currently in the destination array
+            inputOut->prev = inputOut->cur;
+            // Copy current inputs from the polled inputs
+            inputOut->cur = input->cur;
+            // Calculate press and rel from these
+            buttonDiff = inputOut->prev.button ^ inputOut->cur.button;
+            inputOut->press.button = inputOut->cur.button & buttonDiff;
+            inputOut->rel.button = inputOut->prev.button & buttonDiff;
+            PadUtils_UpdateRelXY(inputOut);
+        }
+        input++;
+        inputOut++;
+    }
+}
+
+/**
+ * Fetches the most recently polled inputs from padmgr.
+ *
+ * @param inputs      Array of Input of length MAXCONTROLLERS to copy inputs into
+ * @param gameRequest True if polling inputs for updating the game state
+ */
+void PadMgr_GetInput(Input* inputs, s32 gameRequest) {
+    PadMgr_LockPadData();
+    PadMgr_GetInputNoLock(inputs, gameRequest);
+    PadMgr_UnlockPadData();
+}
+
+/**
+ * Copy of `PadMgr_GetInput`
+ *
+ * @see PadMgr_GetInput
+ */
+void PadMgr_GetInput2(Input* inputs, s32 gameRequest) {
+    PadMgr_LockPadData();
+    PadMgr_GetInputNoLock(inputs, gameRequest);
+    PadMgr_UnlockPadData();
+}
+
+void PadMgr_ThreadEntry() {
+    s16* interruptMsg = NULL;
+    s32 actionBits;
+    s32 exit;
+
+    osCreateMesgQueue(&sPadMgrInstance->interruptQueue, sPadMgrInstance->interruptMsgBuf,
+                      ARRAY_COUNT(sPadMgrInstance->interruptMsgBuf));
+    IrqMgr_AddClient(sPadMgrInstance->irqMgr, &sPadMgrInstance->irqClient, &sPadMgrInstance->interruptQueue);
+
+    actionBits = 0;
+    exit = false;
+
+    while (!exit) {
+        // Process all messages currently in the queue, instead of only a single mssage.
+        // Deduplicates the same message.
+        do {
+            osRecvMesg(&sPadMgrInstance->interruptQueue, (OSMesg*)&interruptMsg, OS_MESG_BLOCK);
+            switch (*interruptMsg) {
+                case OS_SC_RETRACE_MSG:
+                    actionBits |= PADMGR_RETRACE_MSG;
+                    break;
+
+                case OS_SC_PRE_NMI_MSG:
+                    actionBits |= PADMGR_PRE_NMI_MSG;
+                    break;
+
+                case OS_SC_NMI_MSG:
+                    actionBits |= PADMGR_NMI_MSG;
+                    break;
+            }
```

Comment:
Add a default case?

Reply:
I don't think an empty default case is any more clear than omitting it, at least in this case

Reply:
Not having an empty `default` case should be fine, yeah